app/uk/gov/hmrc/tai/controllers/income/IncomeController.scala

```diff
@@ -48,17 +53,72 @@ class IncomeController @Inject()(incomeService: IncomeService,
       } recoverWith taxAccountErrorHandler
   }
 
-  def taxCodeIncomesForYear(nino: Nino, year:TaxYear): Action[AnyContent] = authentication.async {
+  def taxCodeIncomesForYear(nino: Nino, year: TaxYear): Action[AnyContent] = authentication.async {
     implicit request =>
       incomeService.taxCodeIncomes(nino, year).map {
         case Seq() => NotFound
         case taxCodeIncomes => Ok(Json.toJson(ApiResponse(Json.toJson(taxCodeIncomes)(Writes.seq(taxCodeIncomeSourceWrites)), Nil)))
-      }recoverWith taxAccountErrorHandler
+      } recoverWith taxAccountErrorHandler
+  }
+
+  def matchedTaxCodeIncomesForYear(nino: Nino, year: TaxYear, incomeType: String, status: String): Action[AnyContent] = authentication.async {
+    implicit request =>
+
+      def filterMatchingEmploymentsToIncomeSource(employments: Seq[Employment], filteredTaxCodeIncomes: Seq[TaxCodeIncome]): Seq[IncomeSource] =
+        filteredTaxCodeIncomes.flatMap { income =>
+          employments.flatMap(emp =>
+            income.employmentId.fold(Seq.empty[IncomeSource]) {
+              id => if (id == emp.sequenceNumber) Seq(IncomeSource(income, emp)) else Seq.empty[IncomeSource]
+            }
+          )
+        }
+
+      def filterTaxCodeIncomes(taxCodeIncomes: Seq[TaxCodeIncome], incomeStatus: String): Seq[TaxCodeIncome] = {
+        if (incomeStatus == Live.toString) {
+          taxCodeIncomes.filter(income => income.status == Live && income.componentType.toString == incomeType)
+        }
+        else {
+          taxCodeIncomes.filter(income => income.status != Live && income.componentType.toString == incomeType)
+        }
+      }
+
+      (for {
+        taxCodeIncomes <- incomeService.taxCodeIncomes(nino, year)
+        filteredTaxCodeIncomes: Seq[TaxCodeIncome] = filterTaxCodeIncomes(taxCodeIncomes, status)
+        employments: Seq[Employment] <-
+          if (filteredTaxCodeIncomes.isEmpty) {
+            Future.successful(Seq.empty[Employment])
+          }
+          else {
+            employmentService.employments(nino, year)
+          }
+        result: Seq[IncomeSource] = filterMatchingEmploymentsToIncomeSource(employments, filteredTaxCodeIncomes)
+      } yield (result: Seq[IncomeSource]) match {
+        case _ => Ok(Json.toJson(ApiResponse(Json.toJson(result), Nil)))
+      }) recoverWith taxAccountErrorHandler
+  }
+
+  def nonMatchingCeasedEmployments(nino: Nino, year: TaxYear): Action[AnyContent] = authentication.async {
+    implicit request =>
+
+      def filterNonMatchingCeasedEmploymentsWithEndDate(employments: Seq[Employment], taxCodeIncomes: Seq[TaxCodeIncome]): Seq[Employment] =
+        employments
+          .filter(emp => !taxCodeIncomes.exists(tci => tci.employmentId.contains(emp.sequenceNumber)))
+          .filter(_.endDate.isDefined)
+
+      (for {
+        taxCodeIncomes <- incomeService.taxCodeIncomes(nino, year)
```

Comment:
Don't need a for comprehension here to resolve non futures, you could pull them out and make it clear why the for / yield block is being used